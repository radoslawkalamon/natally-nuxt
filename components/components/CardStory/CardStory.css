.card-story {
  background-color: var(--color-betwixt);
  border-radius: var(--border-radius);
  box-shadow: inset 0 0 0 var(--border-size) var(--color-shadow), 0 var(--border-size) 0 var(--color-shadow);
  box-sizing: border-box;
  color: var(--color-white);
  display: block;
  height: 200px;
  max-width: 400px;
  overflow: hidden;
  position: relative;
  text-decoration: none;
  transition: box-shadow var(--time-animation) ease, opacity var(--time-animation) ease;
  width: 100%;
}

.card-story:hover {
  opacity: 0.9;
  text-decoration: underline;
}

.card-story:active {
  box-shadow: inset 0 0 0 var(--border-size) var(--color-shadow), 0 0 0 var(--color-shadow);
  opacity: 0.75;
}

.card-story__cover {
  box-shadow: inset 0 0 0 var(--border-size) var(--color-shadow);
  box-sizing: border-box;
  height: 100%;
  left: 0;
  object-fit: cover;
  position: absolute;
  top: 0;
  width: 100%;
  z-index: 1;
}

.card-story__content {
  background: linear-gradient(0deg, rgba(0, 0, 0, 1) 0%, rgba(0, 0, 0, 0) 100%);
  bottom: 0;
  box-sizing: border-box;
  display: grid;
  gap: var(--gap);
  grid-template:
    'date-published reading-time audiobook'
    'title title title';
  left: 0;
  padding: 3em 1em 1em;
  position: absolute;
  width: 100%;
  z-index: 1;
}

.card-story__date-published {
  grid-area: date-published;
  text-align: left;
}

.card-story__reading-time {
  grid-area: reading-time;
  text-align: center;
}

.card-story__audiobook {
  grid-area: audiobook;
  text-align: right;
}

.card-story__audiobook--disabled {
  opacity: 0.33;
  text-decoration: line-through;
}

.card-story__title {
  grid-area: title;
  margin: 0 !important;
}

.card-story:hover .card-story__title {
  text-decoration: underline;
}
