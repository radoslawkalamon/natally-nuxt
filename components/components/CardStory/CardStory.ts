import Vue from 'vue'
import { DTOMetaPostStory } from '@/utils/dto.meta.post.story'
import { formatterDateMeta } from '@/utils/formatter.date.meta'
import { formatterTimeReading } from '@/utils/formatter.time.reading'

export default Vue.extend({
  name: 'ComponentsCardStory',
  props: {
    metaPostStory: {
      type: DTOMetaPostStory,
      required: true
    }
  },
  computed: {
    path (): string {
      return this.metaPostStory.path
    },
    imageCover (): string {
      return this.metaPostStory.imageCover
    },
    imageCover2x (): string {
      return this.metaPostStory.imageCover2x
    },
    datePublished (): string {
      return formatterDateMeta(this.metaPostStory.createdAt)
    },
    datePublishedLabel (): string {
      return `Opublikowano ${this.datePublished}`
    },
    readingTime (): string {
      return formatterTimeReading(this.metaPostStory.timeReading)
    },
    readingTimeLabel (): string {
      return `Tekst przeczytasz w ${this.readingTime}`
    },
    isAudiobook (): boolean {
      return this.metaPostStory.audiobookId !== '0'
    },
    isAudiobookClasses (): string[] {
      return [
        'card-story__audiobook',
        !this.isAudiobook ? 'card-story__audiobook--disabled' : ''
      ]
    },
    isAudiobookLabel (): string {
      return this.isAudiobook
        ? 'Post posiada wersję dźwiękową'
        : 'Post nie posiada wersji dźwiękowej'
    },
    title (): string {
      return this.metaPostStory.title
    }
  }
})
