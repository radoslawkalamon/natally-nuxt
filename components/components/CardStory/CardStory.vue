<template>
  <article
    class="card-story"
    data-test="components-card-story"
  >
    <NuxtLink
      class="card-story__link"
      :to="path"
    >
      <ComponentsCover
        class="card-story__cover"
        :images="[imageCover, imageCover2x]"
      />
      <div class="card-story__content">
        <footer class="card-story__footer">
          <ComponentsTextMeta
            :aria-label="datePublishedLabel"
            class="card-story__date-published"
            :text="datePublished"
            :title="datePublishedLabel"
          />
          <ComponentsTextMeta
            :aria-label="readingTimeLabel"
            class="card-story__reading-time"
            :text="readingTime"
            :title="readingTimeLabel"
          />
          <ComponentsTextMeta
            :aria-label="isAudiobookLabel"
            :class="isAudiobookClasses"
            data-test="components-card-story-audiobook"
            text="Audiobook"
            :title="isAudiobookLabel"
          />
        </footer>
        <header class="card-story__header">
          <ComponentsTitle
            class="card-story__title"
            :design="3"
            :title="title"
            :type="1"
            :shall-show-underscore="false"
          />
        </header>
      </div>
    </NuxtLink>
  </article>
</template>

<script src="./CardStory.ts" lang="ts"></script>
<style src="./CardStory.css" scoped></style>
