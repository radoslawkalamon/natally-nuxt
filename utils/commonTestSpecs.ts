import Vue from 'vue'
import { mount } from '@vue/test-utils'
import { merge } from 'lodash'

export function shallRender (component: typeof Vue, dataTestID: string, defaultOptions?: object): void {
  test('shall render', (): void => {
    const wrapper = mount(component, defaultOptions)
    const isWrapperVisible = wrapper.find(`[data-test="${dataTestID}"]`).isVisible()
    expect(isWrapperVisible).toBeTruthy()
  })
}

export function shallHaveSlot (component: typeof Vue, slot: string, defaultOptions?: object): void {
  test(`shall have slot '${slot}'`, (): void => {
    const text = `This is slot '${slot}' test.`
    const wrapperOptions = merge(defaultOptions, {
      slots: {
        [slot]: text
      }
    })
    const wrapper = mount(component, wrapperOptions)
    expect(wrapper.html()).toContain(text)
  })
}

export function shallHaveStringProp (component: typeof Vue, propName: string, defaultOptions?: object): void {
  test(`shall have string prop '${propName}'`, (): void => {
    const text = `This is prop ${propName} test.`
    const wrapperOptions = merge(defaultOptions, {
      propsData: {
        [propName]: text
      }
    })
    const wrapper = mount(component, wrapperOptions)
    expect(wrapper.html()).toContain(text)
  })
}

export function shallHavePropInfluenceOverClassNames (
  component: typeof Vue,
  propName: string,
  propValue: unknown,
  classNameSuffix: string,
  defaultOptions?: object
) {
  test(`shall have ${typeof propValue} prop '${propName}' influence over class names`, () => {
    const wrapperOptions = merge(defaultOptions, {
      propsData: {
        [propName]: propValue
      }
    })
    const wrapper = mount(component, wrapperOptions)
    const isWrapperHaveClass = wrapper.classes().some(cls => cls.includes(classNameSuffix))
    expect(isWrapperHaveClass).toBeTruthy()
  })
}
