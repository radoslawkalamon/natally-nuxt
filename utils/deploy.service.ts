/* eslint-disable no-console */
import { Client, FileInfo, FTPResponse } from 'basic-ftp'
import { ProgressInfo } from 'basic-ftp/dist/ProgressTracker'
import capitalize from './formatters/capitalize'
import spacelize from './formatters/spacelize'

type DeployServiceConfig = {
  host: string;
  port: number;
  username: string;
  password: string;
  remoteCatalog: string;
  remoteIgnoreFile: string[];
  localCatalog: string;
}

type RemoveStrategies = {
  [key in FileInfo['type']]: (name: FileInfo['name']) => Promise<void | FTPResponse>
}

function decoratorLogCall () {
  return function (_target: any, propertyKey: string, descriptor: PropertyDescriptor) {
    const targetMethod = descriptor.value

    descriptor.value = function (...args: any[]) {
      const formattedName = spacelize(capitalize(propertyKey))
      console.log(`::: ${formattedName} :::`)
      return targetMethod.apply(this, args)
    }

    return descriptor
  }
}

class DeployService {
  readonly host: string;
  readonly port: number;
  readonly username: string;
  readonly password: string;
  readonly remoteCatalog: string;
  readonly remoteIgnoreFile: string[];
  readonly localCatalog: string;
  FTPClient = new Client()

  constructor (config: DeployServiceConfig) {
    this.host = config.host
    this.port = Number(config.port)
    this.username = config.username
    this.password = config.password
    this.remoteCatalog = config.remoteCatalog
    this.remoteIgnoreFile = config.remoteIgnoreFile
    this.localCatalog = config.localCatalog
  }

  public async start (): Promise<void> {
    try {
      await this.openConnection()
      await this.ensureRemoteCatalog()
      await this.openRemoteCatalog()
      await this.cleanRemoteCatalog()
      await this.uploadToRemoteCatalog()
    } catch (e) {
      console.error(e)
      throw e
    } finally {
      this.closeConnection()
    }
  }

  @decoratorLogCall()
  private async openConnection (): Promise<void> {
    await this.FTPClient.access({
      host: this.host,
      port: this.port,
      user: this.username,
      password: this.password,
      secure: false
    })
  }

  @decoratorLogCall()
  private async ensureRemoteCatalog (): Promise<void> {
    await this.FTPClient.ensureDir(this.remoteCatalog)
  }

  @decoratorLogCall()
  private async openRemoteCatalog (): Promise<void> {
    await this.FTPClient.cd(this.remoteCatalog)
  }

  @decoratorLogCall()
  private async cleanRemoteCatalog (): Promise<void> {
    const remoteFiles = await this.FTPClient.list()

    for (const file of remoteFiles) {
      if (this.remoteIgnoreFile.includes(file.name)) {
        continue
      }

      const removeStrategies: RemoveStrategies = {
        0: async () => await Promise.resolve(),
        1: async name => await this.FTPClient.remove(name),
        2: async name => await this.FTPClient.removeDir(name),
        3: async () => await Promise.resolve()
      }
      await removeStrategies[file.type](file.name)
    }
  }

  @decoratorLogCall()
  private async uploadToRemoteCatalog (): Promise<void> {
    this.FTPClient.trackProgress(this.handlerUploadProgress)
    await this.FTPClient.uploadFromDir(this.localCatalog)
    this.FTPClient.trackProgress()
  }

  private handlerUploadProgress (info: ProgressInfo): void {
    if (info.bytes === 0) {
      process.stdout.write(`Uploading file ${info.name}\n`)
    }
  }

  @decoratorLogCall()
  private closeConnection (): void {
    this.FTPClient.close()
  }
}

export default DeployService
