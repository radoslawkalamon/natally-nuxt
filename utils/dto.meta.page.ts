export type DTOMetaPageConstructor = {
  createdAt: string;
  description: string;
  imageOpenGraph?: string;
  path: string;
  title: string;
  updatedAt: string;
}

const getPath = (path: string) => {
  const pathWithoutIndex = path.split('/').filter(i => i !== 'index').join('/')
  return `${pathWithoutIndex}${pathWithoutIndex.at(-1) !== '/' ? '/' : ''}`
}

export class DTOMetaPage {
  readonly createdAt: Date;
  readonly description: string;
  readonly imageOpenGraph: string;
  readonly path: string;
  readonly title: string;
  readonly updatedAt: Date;

  constructor ({
    createdAt,
    description,
    imageOpenGraph = 'common.opengraph.png',
    title,
    updatedAt,
    path
  }: DTOMetaPageConstructor) {
    this.createdAt = new Date(createdAt)
    this.description = description
    this.imageOpenGraph = `/images/${imageOpenGraph}`
    this.path = getPath(path)
    this.title = title
    this.updatedAt = new Date(updatedAt)
  }
}
