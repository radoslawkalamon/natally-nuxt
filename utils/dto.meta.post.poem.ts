import { DTOMetaPage, DTOMetaPageConstructor } from '@/utils/dto.meta.page'

export type DTOMetaPostPoemConstructor = DTOMetaPageConstructor & {
  imageCover: string
  imageOpenGraph: string
}

export class DTOMetaPostPoem extends DTOMetaPage {
  readonly imageCover: string;
  readonly imageOpenGraph: string;

  constructor ({
    createdAt,
    description,
    imageCover,
    imageOpenGraph,
    path,
    title,
    updatedAt
  }: DTOMetaPostPoemConstructor) {
    super({
      createdAt,
      description,
      path,
      title,
      updatedAt
    })

    this.imageCover = `${this.imagePath}${imageCover}`
    this.imageOpenGraph = `${this.imagePath}${imageOpenGraph}`
  }

  get imagePath (): string {
    return '/images/poezja-314/'
  }
}
