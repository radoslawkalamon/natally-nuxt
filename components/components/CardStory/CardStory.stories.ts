import { Meta, Story } from '@storybook/vue'
import CardStory from './CardStory.vue'
import { DTOMetaPostStory } from '@/utils/dto.meta.post.story'

const meta: Meta = {
  title: 'Components / Card Story',
  component: CardStory
}
export default meta

const Template: Story = (_args, { argTypes }) => ({
  components: { CardStory },
  props: Object.keys(argTypes),
  template: '<CardStory v-bind="$props" />'
})

export const Audiobook: Story = Template.bind({})
Audiobook.args = {
  metaPostStory: new DTOMetaPostStory({
    audiobookId: '123456',
    createdAt: '2021-11-01T00:00:00+00:00',
    description: 'Story description',
    imageCover: 'https://www.fillmurray.com/600/350',
    imageCover2x: 'https://www.fillmurray.com/1200/600',
    imageOpenGraph: 'https://www.fillmurray.com/1200/630',
    path: '/poezja-314/hello-world',
    timeReading: 5,
    title: 'Hello world',
    updatedAt: '2021-11-01T00:00:00+00:00'
  })
}

export const NoAudiobook: Story = Template.bind({})
NoAudiobook.args = {
  metaPostStory: new DTOMetaPostStory({
    createdAt: '2021-11-01T00:00:00+00:00',
    description: 'Story description',
    imageCover: 'https://www.fillmurray.com/600/350',
    imageCover2x: 'https://www.fillmurray.com/1200/600',
    imageOpenGraph: 'https://www.fillmurray.com/1200/630',
    path: '/poezja-314/hello-world',
    timeReading: 5,
    title: 'Hello world',
    updatedAt: '2021-11-01T00:00:00+00:00'
  })
}
