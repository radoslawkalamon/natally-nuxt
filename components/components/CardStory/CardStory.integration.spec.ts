import CardStory from '@/components/components/CardStory/CardStory.vue'
import { shallPassIntegrationSanityTest } from '@/devtools/jest.common.spec'
import { createDefaultOptionsFactory } from '@/devtools/jest.common.spec.utils'
import { JestMockIntersectionObserver } from '@/devtools/jest.mock.IntersectionObserver'
import { DTOMetaPostStory } from '@/utils/dto.meta.post.story'

JestMockIntersectionObserver()

const defaultDTOMetaPostStory = {
  createdAt: '2021-11-01T00:00:00+00:00',
  description: 'Test description',
  imageCover: 'image-cover.webp',
  imageCover2x: 'image-cover.2x.webp',
  imageOpenGraph: 'image-open-graph.webp',
  path: '/test-url',
  timeReading: 2,
  title: 'Test title',
  updatedAt: '2021-11-01T00:00:00+00:00'
}

const defaultOptionsFactory = createDefaultOptionsFactory({
  propsData: {
    metaPostStory: new DTOMetaPostStory(defaultDTOMetaPostStory)
  }
})

describe('Components / Card Story', () => {
  describe('Integration', () => {
    shallPassIntegrationSanityTest({
      component: CardStory,
      description: 'no audiobook / not in the viewport',
      options: defaultOptionsFactory()
    })

    shallPassIntegrationSanityTest({
      component: CardStory,
      description: 'audiobook / in the viewport',
      options: defaultOptionsFactory({
        propsData: {
          metaPostStory: new DTOMetaPostStory({
            ...defaultDTOMetaPostStory,
            audiobookId: '123456'
          })
        },
        data () {
          return {
            'common/intersectionObserver/shownOnce': true
          }
        }
      })
    })
  })
})
