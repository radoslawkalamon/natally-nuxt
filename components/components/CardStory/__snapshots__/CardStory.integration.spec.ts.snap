// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components / Card Story Integration shall pass integration sanity test | audiobook / in the viewport 1`] = `
"<article class="card-story">
  <nuxtlink-stub to="/test-url/" class="card-story__link">
    <div aria-hidden="true" class="cover card-story__cover"></div>
    <div class="card-story__content">
      <footer class="card-story__footer">
        <p class="text-meta card-story__date-published" aria-label="Opublikowano 01.11.2021" title="Opublikowano 01.11.2021">01.11.2021</p>
        <p class="text-meta card-story__reading-time" aria-label="Tekst przeczytasz w 2 minuty" title="Tekst przeczytasz w 2 minuty">2 minuty</p>
        <p class="text-meta card-story__audiobook" aria-label="Post posiada wersję dźwiękową" data-test="components-card-story-audiobook" title="Post posiada wersję dźwiękową">Audiobook</p>
      </footer>
      <header class="card-story__header">
        <h1 class="card-story__title title title--design-3">Test title</h1>
      </header>
    </div>
  </nuxtlink-stub>
</article>"
`;

exports[`Components / Card Story Integration shall pass integration sanity test | no audiobook / not in the viewport 1`] = `
"<article class="card-story">
  <nuxtlink-stub to="/test-url/" class="card-story__link">
    <div aria-hidden="true" class="cover card-story__cover"></div>
    <div class="card-story__content">
      <footer class="card-story__footer">
        <p class="text-meta card-story__date-published" aria-label="Opublikowano 01.11.2021" title="Opublikowano 01.11.2021">01.11.2021</p>
        <p class="text-meta card-story__reading-time" aria-label="Tekst przeczytasz w 2 minuty" title="Tekst przeczytasz w 2 minuty">2 minuty</p>
        <p class="text-meta card-story__audiobook card-story__audiobook--disabled" aria-label="Post nie posiada wersji dźwiękowej" data-test="components-card-story-audiobook" title="Post nie posiada wersji dźwiękowej">Audiobook</p>
      </footer>
      <header class="card-story__header">
        <h1 class="card-story__title title title--design-3">Test title</h1>
      </header>
    </div>
  </nuxtlink-stub>
</article>"
`;
